Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/DelegateActionListener.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) Microsoft Corporation
+ *
+ * All rights reserved.
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+ * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+ * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or substantial portions of
+ * the Software.
+ *
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+ * THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.microsoft.tooling.msservices.serviceexplorer;
+
+import com.google.common.base.Preconditions;
+import com.microsoft.azure.toolkit.lib.common.task.AzureTask;
+import com.microsoft.azure.toolkit.lib.common.task.AzureTaskManager;
+import com.microsoft.azuretools.azurecommons.helpers.AzureCmdException;
+import com.microsoft.azuretools.azurecommons.helpers.NotNull;
+import com.microsoft.azuretools.telemetrywrapper.EventType;
+import com.microsoft.azuretools.telemetrywrapper.EventUtil;
+import com.microsoft.azuretools.telemetrywrapper.Operation;
+import com.microsoft.azuretools.telemetrywrapper.TelemetryManager;
+import com.microsoft.tooling.msservices.components.DefaultLoader;
+import lombok.Getter;
+import lombok.Lombok;
+
+class DelegateActionListener extends NodeActionListener {
+
+    protected final NodeActionListener delegate;
+
+    private DelegateActionListener(NodeActionListener delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void actionPerformed(NodeActionEvent e) throws AzureCmdException {
+        delegate.actionPerformed(e);
+    }
+
+    @Override
+    public AzureIconSymbol getIconSymbol() {
+        return delegate.getIconSymbol();
+    }
+
+    @Override
+    public int getPriority() {
+        return delegate.getPriority();
+    }
+
+    @Override
+    public int getGroup() {
+        return delegate.getGroup();
+    }
+
+    static final class BackgroundActionListener extends DelegateActionListener {
+
+        private final String progressMessage;
+        private boolean cancellable;
+        private boolean conditional;
```

Comment:
conditional is not proper, foreground?

Reply:
I feel conditional is more easy-understand. and it is referenced from com.intellij.openapi.progress.Task.ConditionalModal.
The meaning of "conditional" is keeping our background progress with a dialog condtionally or not. that can be switch between foreground and background, in fact, it's not a foreground.